pkg/networkpolicy/controller.go

```diff
@@ -695,6 +695,14 @@ func (c *Controller) syncNFTablesRules(ctx context.Context) error {
 			Rule: knftables.Concat(
 				"icmpv6", "type", "{", "nd-neighbor-solicit, nd-neighbor-advert", "}", "accept"),
 		})
+		// Don't process traffic generated from the Node, as it can block kubelet probes
+		// or system daemons that depend on the internal node traffic to not be blocked.
+		// Ref: https://github.com/kubernetes-sigs/kube-network-policies/issues/65
+		tx.Add(&knftables.Rule{
+			Chain: chainName,
+			Rule: knftables.Concat(
+				"fib", "saddr", "type", "local", "accept"),
```

Comment:
(nit: you don't need to use `knftables.Concat` here; you can just write `"fib saddr type local accept"`)

But that's not right; with some network plugins, all local pod IPs would be considered "type local" (because the IPs are visible on interfaces in the host netns).

I think the "standard" approach is to allow only from the primary and secondary node IP (learned either via the downward API, or by watching the Node object). Though actually, I don't think that's CNI-agnostic; in some cases, node-to-pod traffic will use the IP of the bridge that the pods are on (eg `docker0`) rather than the official node IP.

The super-clever option would be to figure out kubelet's cgroup ID and write a `meta cgroupv2 ...` rule. Although I think some people expect the non-clever "allow from the node's IP" version anyway.

Reply:
what was wrong with the `iif lo` approach?

Reply:
We have contributors willing to accept challenges , looking at @uablrek now :) 

I like the idea of restricting the loophole to the kubelet, now that you explain it the "type local" sounds easy to cheat

Reply:
> what was wrong with the iif lo approach?

the trace does not show the iif metadata on the postrouting hook, so I think that to make that work we should add a new rule in INPUT to mark the packets and then in POSTROUTING ignore those packets

> cgroups

```
cat /proc/66685/cgroup
0::/kubelet.slice/kubelet.service
```

I don't know how stable or configurable is this option, @BenTheElder do you know?

There is also a skuid and skgid match
https://wiki.nftables.org/wiki-nftables/index.php/Matching_packet_metainformation

```
cat /proc/66685/status
Name:   kubelet
Umask:  0022
State:  S (sleeping)
Tgid:   66685
Ngid:   0
Pid:    66685
PPid:   1
TracerPid:      0
Uid:    0       0       0       0
Gid:    0       0       0       0
```

Maybe accepting everything that comes as root from the host is the best approach, kubelet hast to run as root and if you are root does not matter network policies, you can do whatever you want and access the pods directly on the namespaces

```
        chain trace_chain2 {
                type filter hook postrouting priority filter - 1; policy accept;
                meta skuid 0 counter packets 194 bytes 20003

```

Reply:
I think you should go back to "type local". It's probably not "more unsafe" than anything existing, and to assume user 0 for all traffic I think will cause problems. People try very hard to _not_ enforce the root user it seems.

Reply:
Ok, maybe not "type local", but I think it's unsafe to assume that kubelet runs as root in all clusters, and will so forever

Reply:
But then again, I have no better idea.

Reply:
kubelet has to do root operations creating cgroups per example, the probes need to open sockets ... it can be rootless https://kubernetes.io/docs/tasks/administer-cluster/kubelet-in-userns/ but in a rootless environment this project will not work too 😄 , the network there is 🕺 


> The network namespace of the Node components has to have a non-loopback interface, which can be for example configured with [slirp4netns](https://github.com/rootless-containers/slirp4netns), [VPNKit](https://github.com/moby/vpnkit), or [lxc-user-nic(1)](https://www.man7.org/linux/man-pages/man1/lxc-user-nic.1.html).
> 
> The network namespaces of the Pods can be configured with regular CNI plugins. For multi-node networking, Flannel (VXLAN, 8472/UDP) is known to work.
> 
> Ports such as the kubelet port (10250/TCP) and NodePort service ports have to be exposed from the Node network namespace to the host with an external port forwarder, such as RootlessKit, slirp4netns, or [socat(1)](https://linux.die.net/man/1/socat).

@BenTheElder  how safe do you think it is assume that kubelet runs as root?